### ISSUE TITLE
Feeding non-supported video capability to AddLocalVideoTrackAsync does not generate local video

### ISSUE THREAD

**Issue by eirikhollis** [microsoft/MixedReality-WebRTC#86]
When initializing local video stream with `AddLocalVideoTrackAsync( LocalVideoTrackSettings )`, using width/height/framerate that is not supported by the camera will cause the local video to not be intialized.
 
Expected behaviour is either default values (easy) or a closest-to-given-settings approach (less easy) which would choose the capability closest to the one given to  `AddLocalVideoTrackAsync`.

**Comment by djee-ms** [comment]
This is by design. Arguable maybe, and I'm open to discussion, but by design.

The rationale is that either you look for simplicity and let WebRTC decide for you, which I expect is the common use case for most users, or you want to take full control but then you need to list the supported resolutions for the video capture device you selected and pick one. `LocalVideoTrackSettings` is not a media constraints mechanism, it is intended to force one very specific device, and optionally one specific capture format for that device. And you can build your own "closest-to-given-settings" algorithm if you want to, customized to your own needs. But I prefer to have initializing fail if you request a resolution that doesn't exist, to make it clear something is wrong.

I think you are used to the Google implementation, which has in my opinion a very different target audience (Chromium users) where their users are not developers, and so they need to get a capture device open at all cost, even if it means ignoring any single constraint passed by the user; realistically most Skype users (as a familiar example of video chat; but think also Stadia if that's more believable) do not care what resolution is used as long as "it looks good". The number one priority is to get something, anything. I assume on the other hand that the audience for MixedReality-WebRTC are developers, which when asking for a particular format expect the code to use that format or to return an error, and not go decide with an arbitrary algorithm to change their ask. And then we provide the tools to list the available formats, so that a developer can build an adaptive algorithm if its use case calls for it. But at least they know exactly what is going on. Because I have seen some discrepancies in Google's "best-fit" algorithm depending on where it's called (_e.g._ sometimes it ignores the framerate, sometimes not), and that in my opinion is more confusing than anything.

**Comment by djee-ms** [comment]
@eirikhollis if that's OK with you I am going to close that issue as Won't Fix. I still believe that putting the decision in the hands of the developer is better, and now that #90 is fixed there should be nothing preventing a developer from crafting a custom algorithm suited to their use, whether it's a best fit or some default value or a hard fail, the latter being relevant for _e.g._ HoloLens 1/2 where the hardware is fixed and known in advance. Let me know if you have any other concern about this, or proposition to improve things; maybe rename some variables around width/height/framerate constraints?

**Comment by eirikhollis** [comment]
Sorry I haven't responded. I agree with your explanation and decision.
Only other discussion I can provide is what you think about changing width/height/framerate to a predefined video profile id such as the webrtc-uwp-sdk implementation does it.
I'm not entirely sure if that's more intuitive and would lile your opinion on it. 
You may close this issue.

**Comment by djee-ms** [comment]
You can already pass in order of control and strictness:
1. a video profile unique ID, which selects that one profile
2. a video profile kind, which selects a group of profiles
3. nothing, which selects all available profiles

And in parallel you can add some optional width/height/framerate to select a particular capture format inside the selected profile (case 1.), or if multiple profiles are available (case 2. and 3.) find the first profile with a capture format that matches those criteria.

What do you mean by "changing width/height/framerate to a predefined video profile ID"?

**Comment by eirikhollis** [comment]
Thinking about the capturecapability object which holds an Id, Width, Height and Framerate, such that you pass a single object instead of multiple values.
But nevermind that, seems as if there are plenty of options readily available 👍

**Comment by djee-ms** [comment]
Ok closing this one then, thanks.
